File: Stack/Parenthesis_Checker.java
class Parenthesis_Checker{

    static boolean ispar(String x)
    {
        // Using Stack
        // Time complexity O(n) || Space complexity O(n)
        int len = x.length();
        
        if(len % 2 != 0){
            return false;
        }
        
        Stack<Character> st = new Stack<>();
        
        for(int i=0; i<len; i++){
            
            char ch = x.charAt(i);
            if(ch == '{' || ch == '(' || ch == '['){
                st.push(ch);
            }
            else{
                if(st.isEmpty()){
                    return false;
                }
                
                char top = st.peek();
                if(!((top == '{' && ch == '}') || (top == '(' && ch == ')') || (top == '[' && ch == ']'))){
                    return false;
                }
                st.pop();
            }
            
        }
        
        if(st.isEmpty()){
            return true;
        }
        
        return false;
    }

    public static void main(String args[]){

    }
}

File: Stack/Parenthesis_Checker.cpp
#include <iostream>

using namespace std;

// Using a single loop
// Time complexity O(n) || Space complexity O(1)
class Solution
{
    public:
    //Function to check if brackets are balanced or not.
    bool ispar(string x)
    {
        int len = x.length();
        int i = -1;
        
        for(int j=0; j<len; j++){
            char ch = x[j];
            
            if(ch == '(' || ch == '{' || ch == '['){
                i++;
                x[i] = ch;
            }
            else{
                if(i>=0 && ((x[i]=='(' && ch==')') || (x[i]=='{' && ch=='}') || (x[i]=='[' && ch==']'))){
                    i--;
                }
                else{
                    return false;
                }
            }
        }
        
        if(i == -1){
            return true;
        }
        return false;
    }

};